### PR TITLE
Feature/2505 enable tagging users in comments

### LINKED ISSUE #2505: [API view] Enable tagging users in comments
When adding comments to APIs, enable tagging users with `@username` and send them an email notification.

### PR DESCRIPTION
**Pull request by Vailorium**
Adds user tagging feature to review comments and replies
image

- Grabs users from application's github required organisations
- Suggestion box for tagging pops up after typing @ + 2 characters
- Suggestions are sorted a-z, profile image is included, max 5 suggestions at once (can be adjusted) (see above)
- Suggestions can be entered by clicking, or using arrow keys + tab/enter

Email notifications (TODOs) will be handled with #4164 as it is not reasonable to complete at this point.

Resolves #2505

### CODE REVIEW

File: src/dotnet/APIView/APIViewWeb/Pages/Shared/_CommentFormPartial.cshtml [hunk 1]
```diff
@@ -1,8 +1,19 @@
-﻿<div id="js-comment-form-template" class="d-none">
+﻿@using APIViewWeb.Models
+@model HashSet<GithubUser>
+<div id="js-comment-form-template" class="d-none">
     <div class="comment-form border-top new-thread-comment">
         <form data-post-update="comments" class="new-thread-comment-form comment" method="post" asp-controller="Comments" asp-action="Add">
-            <div class="new-comment-content">
+            <div class="new-comment-content position-relative">
```
Comment (chidozieononiwu) [inline src/dotnet/APIView/APIViewWeb/Pages/Shared/_CommentFormPartial.cshtml:6]: Looks like position: relative already exist in site.css for [new-comment-content](https://github.com/Azure/azure-sdk-tools/blob/57547950e8c84df91bfbe134836bb5aab558c74f/src/dotnet/APIView/APIViewWeb/Client/css/site.scss#L1045) is there any reason to also add it here?

File: src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs [hunk 2]
```diff
@@ -47,12 +88,22 @@ public async Task AddCommentAsync(ClaimsPrincipal user, CommentModel comment)
...
             comment.Comment = commentText;
+
+            HashSet<string> newTaggedUsers = new HashSet<string>();
+            foreach(string taggedUser in taggedUsers)
+            {
+                if(!comment.TaggedUsers.Contains(taggedUser))
+                {
+                    // TODO: notify users they have been tagged
+                }
+                newTaggedUsers.Add(taggedUser);
```
Comment (chidozieononiwu) [inline src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs:105]: Should newTaggedUsers be passed into one of the methods below.

**Review by chidozieononiwu — COMMENTED** [review]
Needs Rebase. Few comments otherwise it looks fine to me.

File: src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs [hunk 3]
```diff
@@ -47,12 +88,22 @@ public async Task AddCommentAsync(ClaimsPrincipal user, CommentModel comment)
...
+            }
```
Comment (praveenkuttappan) [inline src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs:106]: My guess is comment.taggedUsers should be updated with additional new users.

File: src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs [hunk 4]
```diff
@@ -47,12 +88,22 @@ public async Task AddCommentAsync(ClaimsPrincipal user, CommentModel comment)
...
 
-        public async Task<CommentModel> UpdateCommentAsync(ClaimsPrincipal user, string reviewId, string commentId, string commentText)
+        public async Task<CommentModel> UpdateCommentAsync(ClaimsPrincipal user, string reviewId, string commentId, string commentText, string[] taggedUsers)
         {
             var comment = await _commentsRepository.GetCommentAsync(reviewId, commentId);
             await AssertOwnerAsync(user, comment);
             comment.EditedTimeStamp = DateTime.Now;
             comment.Comment = commentText;
+
+            HashSet<string> newTaggedUsers = new HashSet<string>();
```
Comment (praveenkuttappan) [inline src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs:98]: Shouldn't we populate this with current tagged users as initial value and then just add new users into this hashset to create new tagged users?

File: src/dotnet/APIView/APIViewWeb/Pages/Shared/_CommentFormPartial.cshtml (hunk 1 above)
Comment (Vailorium) [reply src/dotnet/APIView/APIViewWeb/Pages/Shared/_CommentFormPartial.cshtml]: Not sure why I added it (must've assumed it didn't have it?). Removed now

File: src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs (hunk 4 above)
Comment (Vailorium) [reply src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs:98]: I did it this way as the comment author might remove tagged users as well as adding them

File: src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs (hunk 3 above)
Comment (Vailorium) [reply src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs:106]: Yes - I planned to finish this when handling the user notification portion, but as that was moved to #4164 I forgot - fixed now!

File: src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs [hunk 5]
```diff
@@ -19,14 +25,48 @@ public class CommentsManager
...
+            IOptions<OrganizationOptions> options)
         {
             _authorizationService = authorizationService;
             _commentsRepository = commentsRepository;
             _notificationManager = notificationManager;
+            _Options = options.Value;
+
+            TaggableUsers = new HashSet<GithubUser>();
+
+            LoadTaggableUsers();
```
Comment (praveenkuttappan) [inline src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs:45]: This is fine for now. Only catch is that it is populated only during startup of server. APIView uses singleton class that's created at startup. This is not a blocker for now and we can improvise it later.

**Review by praveenkuttappan — APPROVED** [review]

File: src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs (hunk 5 above)
Comment (maririos) [reply src/dotnet/APIView/APIViewWeb/Repositories/CommentsManager.cs:45]: you mean improve? :)

**Pull request merged**